rustls/src/crypto/ring/tls13.rs

```diff
@@ -0,0 +1,224 @@
+
+/// The TLS1.3 ciphersuite TLS_CHACHA20_POLY1305_SHA256
+pub static TLS13_CHACHA20_POLY1305_SHA256: SupportedCipherSuite =
+    SupportedCipherSuite::Tls13(TLS13_CHACHA20_POLY1305_SHA256_INTERNAL);
+
+pub(crate) static TLS13_CHACHA20_POLY1305_SHA256_INTERNAL: &Tls13CipherSuite = &Tls13CipherSuite {
+    common: CipherSuiteCommon {
+        suite: CipherSuite::TLS13_CHACHA20_POLY1305_SHA256,
+        bulk: BulkAlgorithm::Chacha20Poly1305,
+        hash_provider: &super::hash::SHA256,
+    },
+    hmac_provider: &super::hmac::HMAC_SHA256,
+    aead_alg: &AeadChacha20Poly1305(RingAeadFactory(&ring::aead::CHACHA20_POLY1305)),
+    #[cfg(feature = "quic")]
+    confidentiality_limit: u64::MAX,
+    #[cfg(feature = "quic")]
+    integrity_limit: 1 << 36,
+};
+
+/// The TLS1.3 ciphersuite TLS_AES_256_GCM_SHA384
+pub static TLS13_AES_256_GCM_SHA384: SupportedCipherSuite =
+    SupportedCipherSuite::Tls13(&Tls13CipherSuite {
+        common: CipherSuiteCommon {
+            suite: CipherSuite::TLS13_AES_256_GCM_SHA384,
+            bulk: BulkAlgorithm::Aes256Gcm,
+            hash_provider: &super::hash::SHA384,
+        },
+        hmac_provider: &super::hmac::HMAC_SHA384,
+        aead_alg: &AeadAes256Gcm(RingAeadFactory(&ring::aead::AES_256_GCM)),
+        #[cfg(feature = "quic")]
+        confidentiality_limit: 1 << 23,
+        #[cfg(feature = "quic")]
+        integrity_limit: 1 << 52,
+    });
+
+/// The TLS1.3 ciphersuite TLS_AES_128_GCM_SHA256
+pub static TLS13_AES_128_GCM_SHA256: SupportedCipherSuite =
+    SupportedCipherSuite::Tls13(TLS13_AES_128_GCM_SHA256_INTERNAL);
+
+pub(crate) static TLS13_AES_128_GCM_SHA256_INTERNAL: &Tls13CipherSuite = &Tls13CipherSuite {
+    common: CipherSuiteCommon {
+        suite: CipherSuite::TLS13_AES_128_GCM_SHA256,
+        bulk: BulkAlgorithm::Aes128Gcm,
+        hash_provider: &super::hash::SHA256,
+    },
+    hmac_provider: &super::hmac::HMAC_SHA256,
+    aead_alg: &AeadAes128Gcm(RingAeadFactory(&ring::aead::AES_128_GCM)),
+    #[cfg(feature = "quic")]
+    confidentiality_limit: 1 << 23,
+    #[cfg(feature = "quic")]
+    integrity_limit: 1 << 52,
+};
+
+struct Tls13MessageEncrypter {
+    enc_key: ring::aead::LessSafeKey,
+    iv: Iv,
+}
+
+struct Tls13MessageDecrypter {
+    dec_key: ring::aead::LessSafeKey,
+    iv: Iv,
+}
+
+impl MessageEncrypter for Tls13MessageEncrypter {
+    fn encrypt(&self, msg: BorrowedPlainMessage, seq: u64) -> Result<OpaqueMessage, Error> {
+        let total_len = msg.payload.len() + 1 + self.enc_key.algorithm().tag_len();
+        let mut payload = Vec::with_capacity(total_len);
+        payload.extend_from_slice(msg.payload);
+        msg.typ.encode(&mut payload);
+
+        let nonce = aead::Nonce::assume_unique_for_key(make_nonce(&self.iv, seq));
+        let aad = aead::Aad::from(make_tls13_aad(total_len));
+
+        self.enc_key
+            .seal_in_place_append_tag(nonce, aad, &mut payload)
+            .map_err(|_| Error::General("encrypt failed".to_string()))?;
```

Comment:
I think this should be mapping to [`Error::EncryptError`](https://github.com/rustls/rustls/blob/6bdaf04e7af97821485da115b5436913fedea874/rustls/src/error.rs#L48-L50)?

Reply:
good call! added a commit which does this